programs/bpf_loader/src/helpers.rs

```diff
@@ -255,13 +293,466 @@ impl HelperObject<BPFError> for HelperSolAllocFree {
     }
 }
 
+// Cross-program invocation helpers
+
+pub type TranslatedAccounts<'a> = (Vec<Rc<RefCell<Account>>>, Vec<(&'a mut u64, &'a mut [u8])>);
+
+/// Implemented by language specific data structure translators
+trait HelperProcessInstruction<'a> {
+    fn get_context_mut(&self) -> Result<RefMut<&'a mut dyn InvokeContext>, EbpfError<BPFError>>;
+    fn translate_instruction(
+        &self,
+        addr: u64,
+        ro_regions: &[MemoryRegion],
+    ) -> Result<Instruction, EbpfError<BPFError>>;
+    fn translate_accounts(
+        &self,
+        message: &Message,
+        account_infos_addr: u64,
+        account_infos_len: usize,
+        ro_regions: &[MemoryRegion],
+        rw_regions: &[MemoryRegion],
+    ) -> Result<TranslatedAccounts<'a>, EbpfError<BPFError>>;
+    fn translate_signers(
+        &self,
+        program_id: &Pubkey,
+        signers_seeds_addr: u64,
+        signers_seeds_len: usize,
+        ro_regions: &[MemoryRegion],
+    ) -> Result<Option<Vec<Pubkey>>, EbpfError<BPFError>>;
+}
+
+/// Cross-program invocation called from Rust
+pub struct HelperProcessInstructionRust<'a> {
+    invoke_context: Rc<RefCell<&'a mut dyn InvokeContext>>,
+}
+impl<'a> HelperProcessInstruction<'a> for HelperProcessInstructionRust<'a> {
+    fn get_context_mut(&self) -> Result<RefMut<&'a mut dyn InvokeContext>, EbpfError<BPFError>> {
+        self.invoke_context
+            .try_borrow_mut()
+            .map_err(|_| HelperError::InvokeContextBorrorFailed.into())
+    }
+    fn translate_instruction(
+        &self,
+        addr: u64,
+        ro_regions: &[MemoryRegion],
+    ) -> Result<Instruction, EbpfError<BPFError>> {
+        let ix = translate_type!(Instruction, addr, ro_regions)?;
+        let accounts = translate_slice!(
+            AccountMeta,
+            ix.accounts.as_ptr(),
+            ix.accounts.len(),
+            ro_regions
+        )?
+        .to_vec();
+        let data = translate_slice!(u8, ix.data.as_ptr(), ix.data.len(), ro_regions)?.to_vec();
+        Ok(Instruction {
+            program_id: ix.program_id,
+            accounts,
+            data,
+        })
+    }
+
+    fn translate_accounts(
+        &self,
+        message: &Message,
+        account_infos_addr: u64,
+        account_infos_len: usize,
+        ro_regions: &[MemoryRegion],
+        rw_regions: &[MemoryRegion],
+    ) -> Result<TranslatedAccounts<'a>, EbpfError<BPFError>> {
+        let account_infos = if account_infos_len > 0 {
+            translate_slice!(
+                AccountInfo,
+                account_infos_addr,
+                account_infos_len,
+                ro_regions
+            )?
+        } else {
+            &[]
+        };
+
+        // TODO ro vs rw translate given the pointers are passed to another program...
```

Comment:
nit: please file an issue instead of embedding a TODO comment

Reply:
Heh, that was a todo for me only and was intended for me to do a review of the pointer usage, I did but forgot to remove the todo :-)